From 160f09d7e54f7108584eb36d3aa08780a2da0115 Mon Sep 17 00:00:00 2001
From: wanlong <long.wan@spacemit.com>
Date: Fri, 25 Oct 2024 11:10:00 +0800
Subject: [PATCH 1200/1448] wireless: rtl8852bs: fix wireless_dev lockdep
 warning

WARNING: CPU: 2 PID: 208 at drivers/net/wireless/realtek/rtl8852bs/os_dep/linux/ioctl_cfg80211.c:241
 rtw_cfg80211_ch_switch_notify+0x8e/0xb8 [8852bs]
Modules linked in: 8852bs
CPU: 2 PID: 208 Comm: disp_eng_share_ Not tainted 6.6.36+ #141
Hardware name: spacemit k1-x deb1 board (DT)
epc : rtw_cfg80211_ch_switch_notify+0x8e/0xb8 [8852bs]
 ra : rtw_cfg80211_ch_switch_notify+0x48/0xb8 [8852bs]
epc : ffffffff03a3be76 ra : ffffffff03a3be30 sp : ffffffc80985bbf0
 gp : ffffffff83541fe0 tp : ffffffd908e94d40 t0 : ffffffc80985bc18
 t1 : ffffffd908c22960 t2 : 0000000000000000 s0 : ffffffc80985bc50
 s1 : ffffffc808b17000 a0 : 0000000000000001 a1 : 00000000004fa6a0
 a2 : ffffffd90c313768 a3 : ffffffff0427f860 a4 : 0000000000000008
 a5 : 0000000000000001 a6 : 00000000000003e8 a7 : ffffffd908c22988
 s2 : 0000000000000001 s3 : ffffffc808bf3000 s4 : 0000000000000001
 s5 : 0000000000001911 s6 : ffffffc808bf3000 s7 : ffffffc808bf36e1
 s8 : 0000000000000000 s9 : 0000000000000000 s10: 0000000000000000
 s11: 0000000000000000 t3 : ffffffda75234908 t4 : ffffffda75d6efb0
 t5 : 0000000000000001 t6 : ffffffc80985bc38
status: 0000000200000120 badaddr: 0000000000000000 cause: 0000000000000003
[<ffffffff03a3be76>] rtw_cfg80211_ch_switch_notify+0x8e/0xb8 [8852bs]
[<ffffffff03a88416>] start_clnt_join+0xec/0x1ca [8852bs]
[<ffffffff03a77230>] _connect_msg_hdlr+0x644/0xaee [8852bs]
[<ffffffff03b33a56>] cur_req_hdl+0x30/0x3a [8852bs]
[<ffffffff03b343bc>] msg_dispatch+0x238/0x3e4 [8852bs]
[<ffffffff03b37b0a>] dispr_thread_loop_hdl+0x25e/0x2d4 [8852bs]
[<ffffffff03b37b90>] dispr_share_thread_loop_hdl+0x10/0x18 [8852bs]
[<ffffffff03b316e8>] share_thread_hdl+0xa0/0x194 [8852bs]
[<ffffffff80041190>] kthread+0xda/0xf6
[<ffffffff80fb661e>] ret_from_fork+0xe/0x1c

Change-Id: I3046ef8882e7a5f739a183da947c26792663868f
---
 .../realtek/rtl8852bs/os_dep/linux/ioctl_cfg80211.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/realtek/rtl8852bs/os_dep/linux/ioctl_cfg80211.c b/drivers/net/wireless/realtek/rtl8852bs/os_dep/linux/ioctl_cfg80211.c
index 707cd12a97f0..a50bf29733d1 100644
--- a/drivers/net/wireless/realtek/rtl8852bs/os_dep/linux/ioctl_cfg80211.c
+++ b/drivers/net/wireless/realtek/rtl8852bs/os_dep/linux/ioctl_cfg80211.c
@@ -227,6 +227,7 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, struct rtw_chan_def *rtw_chd
 {
 	struct wiphy *wiphy = adapter_to_wiphy(adapter);
 	struct _ADAPTER_LINK *alink = GET_PRIMARY_LINK(adapter);
+	struct wireless_dev *wdev = adapter->pnetdev->ieee80211_ptr;
 	u8 ret = _SUCCESS;
 
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 8, 0))
@@ -238,6 +239,8 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, struct rtw_chan_def *rtw_chd
 
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0))
 	if (started) {
+		mutex_lock(&wdev->mtx);
+		__acquire(&wdev->mtx);
 		#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0))
 		cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, alink->mlmepriv.link_id, 0, false, 0);
 		#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
@@ -253,9 +256,11 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, struct rtw_chan_def *rtw_chd
 		 */
 
 		cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0, false);
-#else
+		#else
 		cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0);
-#endif
+		#endif
+		__release(&wdev->mtx);
+		mutex_unlock(&wdev->mtx);
 		goto exit;
 	}
 #endif
@@ -263,6 +268,8 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, struct rtw_chan_def *rtw_chd
 	if (!rtw_cfg80211_allow_ch_switch_notify(adapter))
 		goto exit;
 
+	mutex_lock(&wdev->mtx);
+	__acquire(&wdev->mtx);
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0))
 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef, alink->mlmepriv.link_id, 0);
 #elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 2))
@@ -270,6 +277,8 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, struct rtw_chan_def *rtw_chd
 #else
 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef);
 #endif
+	__release(&wdev->mtx);
+	mutex_unlock(&wdev->mtx);
 
 #else
 	int freq = rtw_bch2freq(rtw_chdef->band, rtw_chdef->chan);
-- 
2.47.0

